### ISSUE TITLE
pytorch nn.module model 'state_dict' does not support item assignmentPylintE1137:unsupported-assignment-operation

### ISSUE THREAD

**Issue by rocke2020** [pylint-dev/pylint#9321]
### Bug description

```python
import torch
import torch.nn as nn


class Test(nn.Module):
    def __init__(
        self,
        embed_dim=3,
    ) -> None:
        super().__init__()
        self.embedding = torch.nn.Embedding(100, embed_dim)

    def forward(self, x):
        inputs = self.embedding(x)
        return inputs

model = Test()
model_state_dict = model.state_dict()
k = 'embedding.weight'
v = torch.tensor([[1, 2, 3.]])
# type(model_state_dict) <class 'collections.OrderedDict'>
print(model_state_dict.keys())
print(type(model_state_dict))
model_state_dict[k] = v
```

### Configuration

```ini
python 3.10.12
pytorch                   2.1.0 
I test both pylint 2.16.2 and 3.0.3
Both give the wrong warning
for the line 
"
model_state_dict[k] = v
"
'model_state_dict' does not support item assignmentPylintE1137:unsupported-assignment-operation
(variable) model_state_dict: Dict[str, Any]
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
'model_state_dict' does not support item assignmentPylintE1137:unsupported-assignment-operation
(variable) model_state_dict: Dict[str, Any]
```

### Expected behavior

Not assignmentPylintE1137:unsupported-assignment-operation warning

### Pylint version

```shell
python 3.10.12
I test both 
1) pylint 2.16.2 and astroid 2.14.2
2) pylint 3.0.3 and  astroid 3.0.2
```

### OS / Environment

ubuntu 20.04

### Additional dependencies

```python
pytorch                   2.1.0   
pandas                    2.1.1 
numpy                     1.25.2
```
